### DISCUSSION TITLE
翻译服务失败时自动顺延选择及独立样式功能申请

### DISCUSSION THREAD

**Discussion by pbxs666** [immersive-translate/immersive-translate#1880] [Ideas]
### 新功能 | New Feature

1.当前的翻译服务因各种原因失败时，希望能**自动顺延选择下一个翻译服务**。不弹出选择翻译的提示框，例如：
![1](https://github.com/immersive-translate/immersive-translate/assets/47630168/9c56b5f1-df7e-4e54-87eb-9e59df49969f)
2.配合上一条功能，翻译服务列表里需要**增加排序功能来定义优先级**。目前可以对每个翻译服务单独启用，但不能排序来自定义优先级。
3.如果再能搭配上**每个翻译服务独立的样式**，就可以直接识别出当前用的是哪家的翻译，对翻译的质量和偏好能有预期
4.如果当前翻译某段时间内一直失败，那每次翻译时都要从列表第一个开始往后顺延，那翻译的速度就要打折扣了，所以能增加一个**当前翻译请求连续失败几次后临时切换到下一个服务保持一段时间**的功能就更好了

### 为什么您想要添加这些功能 | Why you want to add these features

1.偶尔会弹出其他翻译服务选择项，但十有八九都是谷歌，不知道是不是bug，翻译服务里已经关掉谷歌的开关了，目前不想用谷歌
2.该条建议是为了1.能更好的工作所要搭配使用的
3.有了独立的样式，便能在靠前的翻译服务失败然后自动切换下一个服务后，方便的识别出当前是顺延到哪个翻译服务了（如果连续2个或者多个服务都因各种原因失败），每个翻译服务的质量和偏好都是有区别的，这样能对翻译出的结果有心理的预期和后续的对应处理方案

**Comment by theowenyoung** [comment]
这种错误的情况多吗？ 是不是可以使用一个更加稳定点的翻译服务，比如微软翻译，您可以申请 Azure 的提供的官方翻译 APIKey。

（您描述的逻辑，有点过于复杂，如果实现的话，可能让使用插件的人难以理解。目前的逻辑会比较简单直接。

> 1.偶尔会弹出其他翻译服务选择项，但十有八九都是谷歌，不知道是不是bug，翻译服务里已经关掉谷歌的开关了，目前不想用谷歌

 这个检测默认只检测免费翻译服务，目前只有 谷歌，微软，yandex，腾讯可用，所以就直接检测这几个。

**Reply by pbxs666** [reply]
> 这种错误的情况多吗？ 是不是可以使用一个更加稳定点的翻译服务，比如微软翻译，您可以申请 Azure 的提供的官方翻译 APIKey。
> 
> （您描述的逻辑，有点过于复杂，如果实现的话，可能让使用插件的人难以理解。目前的逻辑会比较简单直接。
> 
> > 1.偶尔会弹出其他翻译服务选择项，但十有八九都是谷歌，不知道是不是bug，翻译服务里已经关掉谷歌的开关了，目前不想用谷歌
> 
> 这个检测默认只检测免费翻译服务，目前只有 谷歌，微软，yandex，腾讯可用，所以就直接检测这几个。

嗯嗯，我明白你的意思。

但是图中实例的就是**微软翻译**失败后的情况（至少它在我这网络环境下不是非常稳定），在我这出现次数还是不少的，总是给我提示切换到**谷歌翻译**，还要手动切换到腾讯或百度的翻译。

该怎么说呢，我复述下我的使用场景吧，我经常用nexusmod这个游戏MOD网页，经常会涉及到游戏专有名词的翻译，国内的翻译腾讯相对好点，百度要逊色一些。但综合来看还是比不上国外的翻译，毕竟翻译的词汇都是外文游戏，还是人家国外本土的准确，国内的有时候连名词都翻不出来。

**Reply by pbxs666** [reply]
谷歌的话，需要改host，配置IP。之前配过，但过段时间会失效。虽然这个很简单，但总的来说是有有效期的，不能一劳永逸，用着还是不爽。

**Reply by pbxs666** [reply]
目前是暂时配置了谷歌的ip，但到底不是永久解决方案呢。

那我先简化下我的需求，重新说下（其他细枝末节的要求先不提了）：
1.能给目前的翻译列表排个序，优先级；
2.当前翻译失败后，自动从翻译列表里按优先级依次尝试别的服务
3.给不同的翻译配置独立的样式，便于区分当前的翻译服务。
没配置时候用全局样式，配置后用专有的。

**Reply by pbxs666** [reply]
其实第3点我感觉可以单独提一个issue，这个可能不止我需要。
举个例子：不同的翻译服务专业领域侧重不同，如果有了独立的样式就可以提醒使用者在不同的使用场景下该换翻译了。
同时也能在自动切换翻译服务后知道当前是哪个翻译（如果我这个自动切换翻译的需求能实现的话）

**Reply by pbxs666** [reply]
有点长有些啰嗦，哈哈。

我说完了，接下来就麻烦开发者看看需求与实现可不可行了

**Reply by theowenyoung** [reply]
我理解您的需求～ 不过这个需求对于已经挺复杂的沉浸式翻译来说，更加复杂了，并不是我们想要走的方向。所以建议您针对这个 case 还是选择一个更加稳定的翻译服务。

**Reply by pbxs666** [reply]
> 我理解您的需求～ 不过这个需求对于已经挺复杂的沉浸式翻译来说，更加复杂了，并不是我们想要走的方向。所以建议您针对这个 case 还是选择一个更加稳定的翻译服务。

好吧，我也能理解。

不过没关系，我暂时先使用谷歌翻译，改了下host文件。

质量确实可以，毕竟老牌翻译。

后面用不了了再考虑吧，大不了再改一下host，哈哈。

**Reply by hbread00** [reply]
在机翻中，谷歌是速度与准确度都较好的，AI翻译中，强的也是ChatGPT和Gemini这些，都是海外翻译服务
所以在能使用时，我一般都是在使用这些服务的，通过代理时使用它们一般也是稳定的，但也有一些不算罕见的情况不能使用，比如：
1. 使用手机，因为费电没有一直开启代理软件，或是出门使用流量，走代理速度很慢
2. 被ban，还没切换节点，或是节点正在切换

遇到这种问题是非常希望能够无感切换的，比如设置国内可用的翻译服务为备用，当无法ping通主翻译时就自动用备用的
只是设置一个备用翻译并用简单的规则切换的话，不会像对网站独立设置那样复杂
